# Student Loan Risk with Deep Learning

In [1]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
from pathlib import Path

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `student-loans.csv` file into a Pandas DataFrame. Review the DataFrame, looking for columns that could eventually define your features and target variables.   

In [2]:
# Read the csv into a Pandas DataFrame
file_path = "https://static.bc-edx.com/ai/ail-v-1-0/m18/lms/datasets/student-loans.csv"
loans_df = pd.read_csv(file_path)

# Review the DataFrame
loans_df.head()

,payment_history,location_parameter,stem_degree_score,gpa_ranking,alumni_success,study_major_code,time_to_completion,finance_workshop_score,cohort_ranking,total_loan_score,financial_aid_score,credit_ranking
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,0
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,0
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,0
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,1
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,0


In [3]:
# Review the data types associated with the columns
loans_df.dtypes

,0
payment_history,float64
location_parameter,float64
stem_degree_score,float64
gpa_ranking,float64
alumni_success,float64
study_major_code,float64
time_to_completion,float64
finance_workshop_score,float64
cohort_ranking,float64
total_loan_score,float64


In [4]:
# Check the credit_ranking value counts
loans_df["credit_ranking"].value_counts()

,count
credit_ranking,
1,855
0,744


### Step 2: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “credit_ranking”. The remaining columns should define the features dataset.

In [5]:
# I used Google Colab to help generate this code.
# prompt: # Define the target set y using the credit_ranking column
# # Display a sample of y

y = loans_df["credit_ranking"].values
y[:5]


array([0, 0, 0, 1, 0])

In [6]:
# I used Google Colab to help generate this code.
# prompt: # Define features set X by selecting all columns but credit_ranking
# # Review the features DataFrame

X = loans_df.drop(columns="credit_ranking")
X.head()


,payment_history,location_parameter,stem_degree_score,gpa_ranking,alumni_success,study_major_code,time_to_completion,finance_workshop_score,cohort_ranking,total_loan_score,financial_aid_score
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4


### Step 3: Split the features and target sets into training and testing datasets.


In [7]:
# I used Google Colab to help generate this code. I also worked with a tutor.
# prompt: # Split the preprocessed data into a training and testing dataset
# # Assign the function a random_state equal to 1

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

# Checking to see the shape of the data and how it was split
print(X_train.shape)
print(X_test.shape)


(1199, 11)
(400, 11)


In [8]:
X_train.head()

,payment_history,location_parameter,stem_degree_score,gpa_ranking,alumni_success,study_major_code,time_to_completion,finance_workshop_score,cohort_ranking,total_loan_score,financial_aid_score
1144,7.3,0.320,0.23,2.3,0.066,35.0,70.0,0.99588,3.43,0.62,10.1
73,8.3,0.675,0.26,2.1,0.084,11.0,43.0,0.99760,3.31,0.53,9.2
446,12.5,0.380,0.60,2.6,0.081,31.0,72.0,0.99960,3.10,0.73,10.5
399,8.7,0.765,0.22,2.3,0.064,9.0,42.0,0.99630,3.10,0.55,9.4
647,8.3,0.845,0.01,2.2,0.070,5.0,14.0,0.99670,3.32,0.58,11.0


### Step 4: Use scikit-learn's `StandardScaler` to scale the features data.

In [9]:
# I used Google Colab to help generate this code.
# prompt: # Create a StandardScaler instance
# # Fit the scaler to the features training dataset
# # Fit the scaler to the features testing dataset

# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


---

## Compile and Evaluate a Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [10]:
# I used Google Colab to help generate this code.
# prompt: # Define the the number of inputs (features) to the model
# # Review the number of features

number_input_features = len(X_train.iloc[0])
print(number_input_features)


11


In [11]:
# I used Google Colab to help generate this code.
# prompt: # Define the number of hidden nodes for the first hidden layer
# # Define the number of hidden nodes for the second hidden layer
# # Define the number of neurons in the output layer

hidden_nodes_layer1 = 6
hidden_nodes_layer2 = 3
number_output_neurons = 1


In [12]:
# I used Google Colab to help generate this code.
# prompt: # Create the Sequential model instance
# # Add the first hidden layer
# # Add the second hidden layer
# # Add the output layer to the model specifying the number of output neurons and activation function

nn = Sequential()

# Hidden layer 1
nn.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Hidden layer 2
nn.add(Dense(units=hidden_nodes_layer2, activation="relu"))


# Output layer
nn.add(Dense(units=1, activation="sigmoid"))


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [13]:
# I used Google Colab to help generate this code.
# prompt: # Display the Sequential model summary

nn.summary()


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 6)                   │              72 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 3)                   │              21 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │               4 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 97 (388.00 B)

 Trainable params: 97 (388.00 B)

 Non-trainable params: 0 (0.00 B)

### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [14]:
# I used Google Colab to help generate this code.
# prompt: # Compile the Sequential model

# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [15]:
# I used Google Colab to help generate this code. I also worked with a tutor in reviewing this code
# prompt: # Compile the Sequential model

# Fit the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)


Epoch 1/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5226 - loss: 0.6849
Epoch 2/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5181 - loss: 0.6845
Epoch 3/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5831 - loss: 0.6751 
Epoch 4/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5988 - loss: 0.6740
Epoch 5/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6759 - loss: 0.6483
Epoch 6/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6406 - loss: 0.6418
Epoch 7/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6837 - loss: 0.6276
Epoch 8/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7073 - loss: 0.6046
Epoch 9/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7103 - loss: 0.5945 
Epoch 10/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7149 - loss: 0.5717
Epoch 11/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6879 - loss: 0.5871
Epoch 12/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7557 - 

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [16]:
# I used Google Colab to help generate this code.
# prompt: # Evaluate the model loss and accuracy metrics using the evaluate method and the test data
# # Display the model loss and accuracy results

# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")


13/13 - 0s - 14ms/step - accuracy: 0.7325 - loss: 0.5418
Loss: 0.5418415665626526, Accuracy: 0.7325000166893005


### Step 4: Save and export your model to a keras file, and name the file `student_loans.keras`.


In [17]:
# I used Google Colab to help generate this code.
# prompt: # Set the model's file path
# # Export your model to a keras file

# Set the model's file path
file_path = Path("student_loans.keras")

# Export your model to a HDF5 file
nn.save(file_path)


---
## Predict Loan Repayment Success by Using your Neural Network Model

### Step 1: Reload your saved model.

In [18]:
# I used Google Colab to help generate this code.
# prompt: # Set the model's file path
# # Load the model to a new object

# Set the model's file path
file_path = Path("student_loans.keras")

# Load the model to a new object
nn_imported = tf.keras.models.load_model(file_path)


### Step 2: Make predictions on the testing data and save the predictions to a DataFrame.

In [19]:
# I used Google Colab to help generate this code.
# prompt: # Make predictions with the test data
# # Display a sample of the predictions

y_pred = nn_imported.predict(X_test_scaled)

predictions_df = pd.DataFrame({"Prediction": y_pred.flatten()})
predictions_df.head()



13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step


,Prediction
0,0.670427
1,0.354534
2,0.822741
3,0.774824
4,0.949519


In [20]:
# I used Google Colab to help generate this code.
# prompt: # Save the predictions to a DataFrame and round the predictions to binary results

predictions_df["Prediction"] = predictions_df["Prediction"].apply(lambda x: 1 if x >= 0.5 else 0)
predictions_df.head()


,Prediction
0,1
1,0
2,1
3,1
4,1


### Step 4: Display a classification report with the y test data and predictions

In [21]:
# I used Google Colab to help generate this code.
# prompt: # Print the classification report with the y test data and predictions

from sklearn.metrics import classification_report

print(classification_report(y_test, predictions_df["Prediction"]))


              precision    recall  f1-score   support

           0       0.69      0.78      0.73       188
           1       0.78      0.69      0.73       212

    accuracy                           0.73       400
   macro avg       0.73      0.73      0.73       400
weighted avg       0.74      0.73      0.73       400



---
## Discuss creating a recommendation system for student loans

Briefly answer the following questions in the space provided:

1. Describe the data that you would need to collect to build a recommendation system to recommend student loan options for students. Explain why this data would be relevant and appropriate.

2. Based on the data you chose to use in this recommendation system, would your model be using collaborative filtering, content-based filtering, or context-based filtering? Justify why the data you selected would be suitable for your choice of filtering method.

3. Describe two real-world challenges that you would take into consideration while building a recommendation system for student loans. Explain why these challenges would be of concern for a student loan recommendation system.

Note:I used the following resources to respond to these questions. Xpert, my teacher Bill during open office hours, google, and my own knowledge and experience.

**1. The data I would need to collect to build a system to recommend student loan options for students would include**: student demographics, e.g.: age, income, educational background, credit score (provides a score between 0 and 850(high) that quantifies how well an individual has handled debt in the past and their current debt to available credit ratio), current level of debt, payment history (successful history of past repayment is a good predictor for current and future repayment), academic performance or GPA (quantifies how well a student has done in completing courses; how likely the student is to complete a degree/ graduate); history of and current financial aid status, loans, grants, scholarships, etc. to better understand past funding and current student debt level. Each of these metrics help form a picture of how well a student might be able to successfully pay off student loan debt. Risk prediction for loan repayment. One piece of data that I think is troublesome that was availalbe in this data set is the location parameter. I am assuming this is the address/ zipcode where the student lives. I think this is highly biased. Historically, individuals who resided at zipcodes where a large number of people of color lived, were often denied loans due to nothing more than their zipcode. Ensuring fairness in loan recommendations needs to be of upmost importance when selecting data/features. https://interactive.wbez.org/2020/banking/disparity/  

**2. Based on the data you chose to use in this recommendation system, would your model be using collaborative filtering, content-based filtering, or context-based filtering? Justify why the data you selected would be suitable for your choice of filtering method.**  Collaborative filtering is a method that makes recommendations based on preferences and behaviors of multiple users. There is user based collaborative filtering, for example when netflix recommends a movie based on what you have watched in the past what other viewers who also watched this movie liked & item based collaborative filtering such as amazon recommending another product to purchase, "these items are often purchased together". Content-based filtering takes into consideration an individuals previous actions and feedback. in regards to student loans, it would be based on the specific individuals record and data as described above. Context-based filtering is based on context such as timing, location or need. In regards to student loans this might include marketing the loans so that the timing is near when the need is greatest. Based on my philisophy and the data I would select, content-based filtering is the model I would choose.


**3. Describe two real-world challenges that you would take into consideration while building a recommendation system for student loans. Explain why these challenges would be of concern for a student loan recommendation system.**  I realize my approach to this question is a little unconventional. I don't believe in student loans, except as a very last resort. Too often students are hampered for many, many years with student loan debt. Every effort should be made to help students fund their education through other means and the student loan recommendation system could be designed to help students rather then hurt them.

"Blacks are disporportionatly at risk for carrying high student loans. Among bachelor’s degree holders, black students are the most likely to borrow federal loans at 76.1%. On average, Black or African American student borrowers owe $25,000 more than White or Caucasian borrowers owe for bachelor’s graduates.
Four years after graduation, 48% of Black student borrowers and 17% of White student borrowers owe more than they initially borrowed.
Among graduate students, 40% of Black students and 22% of White students accumulate debt for graduate school." https://educationdata.org/student-loan-debt-statistics

So, I would want to create a student loan system that helps and doesn't hurt students by leveling the playing field for funding and discourages loans.
1. "High vs middle or low GPA ranking" and 2. "time to completion" could represent some real challenges with students. Students who have high GPA (ranking) may be offered mutiple scholarships, grants and loans - even more than needed to cover the cost of schooling - while others may not recieve any scholarships, requiring them to use only student loans to fund their education, resulting in high student loan debt upon completion of their degree. Some scholarships go unawarded simply because no one applied for the scholarship. Because of this, some institutions have begun automatically reviewing students files and matching them with scholarships and grants that they are eliglble for. Grants and scholarships should be considered before any loans are taken out.

Time to completion is also a big concern with higher education. Educational Institutions are graded or scored according to the "on time" graduation rate of their students. Yet, some students choosing to avoid student loans and the resultant debt, may select to work full time and attend school part time - extending the time to completion from 4 years to 6 years. Students who graduate without debt are much better off in the long run. It would be tempting to push loans on students in an effort to increase the institutions time to completion rating. If there are other ways to fund a students education, besides taking out a loan, it should be considered first. Financial literacy should be a part of any financial aid. Protecting student data is also a real concern.

